BUG: assert_array_almost_equal_nulp longdouble

tylerjereddy opened the pull request:
Fixes #21520

* allow `assert_array_almost_equal_nulp` to properly handle `np.longdouble` + regression test

tylerjereddy committed:
BUG: `assert...nulp` longdouble

Fixes #21520

* allow `assert_array_almost_equal_nulp` to properly
handle `np.longdouble` + regression test

charris commented on numpy/testing/_private/utils.py:
```diff
@@ -1794,6 +1794,9 @@ def integer_repr(x):
         return _integer_repr(x, np.int64, np.int64(-2**63))
+    elif x.dtype == np.longdouble:
+        bits = np.finfo(np.longdouble).bits
+        return _integer_repr(x, np.longdouble, np.longdouble(-2**(bits - 1)))
```
I have my doubts about double-double here, but am going to ignore them. I'm also not convinced that the subtraction in `_integer_repr` will work correctly for non-integer types as the exponent bits will be treated differently. Is this discussed somewhere?

charris replied:
To be more precise, using integers the difference counts ulps, although I would have used offset binary and unsigned integers to handle the negatives. The treatment is IEEE specific. For extended and quad precision, 128 bit integers would be desirable.

seberg replied:
Would something like this make sense to avoid the use of integers as much as possible?:
```
def nulp_diff(x, y):
    nmant = np.finfo(x.dtype).nmant + 1  # nmant excludes implicit bit
    if x < y:
        x, y = y, x
    return int((x - y) * 2.**53 / 2**np.frexp(x)[1])
```

charris replied:
The aim is to count representable floats between two values using the fact that both IEEE floats and integers have the same ordering when interpreted as integers, i.e., the number of float values is the same as the number of integer values even though they represent different numbers. This requires all the bits except for the sign bit, and the sign bit could be used if one is a bit clever. NaNs and Infs do cause trouble in this scheme, so need to be handled separately, although I suspect Infs might fit in, haven't  checked.

tylerjereddy replied:
Can you provide concrete test case examples that will fail here if the machine has sufficient precision? Is the failure worse on this branch because it claims to "support" longdouble now?

charris commented:
I don't think there is any easy way to do this correctly given that we need to handle 80 bit extended, 128 quad precision, and double-double. I'm not even sure there _is_ a correct way for double-double, and the others need 128 bit integer arithmetic. We probably just need a workable estimate for numbers not too far apart, although the value of one ulp can change by a factor of two even between two close numbers.

tylerjereddy commented:
The original error message with `unsupported dtype` is starting to sound more sensible after all. Should we change direction to just expanding that error message a bit rather than considering this a "bug?"

seberg commented:
I guess it's an enhancement to begin with?  I do suspect we can define something that only works with the original type and `frexp`, `np.finfo(dtype).minexp`, `np.finfo(dtype).maxexp`, and `np.finfo(dtype).nmant + 1`.
But it will need to deal with `inf` and `NaN` and signs and doing all of that does also seem pretty tedious, so no idea if it's worth prioritizing.

tylerjereddy commented:
The "Handbook of Floating-Point Arithmetic" suggests that in most cases we should be able to use the quantum of the floating point number. I think Sebastian already mentions the exception noted in the book as well--when `x` is a power of the radix.

Maybe I should reframe this slightly though--the equation for this function is actually pretty straightforward as documented in NumPy:

> abs(x - y) <= nulp * spacing(maximum(abs(x), abs(y)))

If the issue at hand is entirely isolated to `assert_array_almost_equal_nulp`, while `np.spacing` performs well for the various implementations of `longdouble`, the problem may simplify a bit. If `np.spacing` needs a similar "enhancement" for `longdouble` implementations, that may be something I need to think about. I'm assuming that there is no shortcut that reframes the problem as a composition of either `np.spacing` or `np.nextafter()`?

tylerjereddy committed:
TST: PR 22770 revisions

* add some larger value test cases--so
far, none of these are failing on a machine
with 128-bit longdouble

* also, adjust the test to enforce a match
within 2 nulp (the truth)--so far, this passes
as well

tylerjereddy commented:
I expanded the test to include larger `longdouble` values like `2**60` mentioned above, and even larger values beyond that. So far, they all pass the new test, which I've also expanded to enforce the truth (2 nulp separation), which also works for all the values I've tested so far.

So, I guess what might be helpful still is an explicit failure scenario/example for `assert_array_almost_equal_nulp` proper with `longdouble`--I still don't have one I don't think. There are some concerns expressed above with regards to `nulp_diff`, but I don't have an actual failing case for `assert_array_almost_equal_nulp` with the new `longdouble` support so far. Maybe the CI will find one.. or perhaps a reviewer can suggest a machine I need to test on--looks like I have 128 bits for my `longdouble` locally.

tylerjereddy commented:
Hmm, maybe I'll have to resort to `hypothesis` probing to look for the weaknesses, if there are any.

tylerjereddy committed:
TST: PR 22770 revisions

* use a more rigorous `hypothesis`-based test

tylerjereddy commented:
Yeah, that did the trick locally at least--`hypothesis` wasn't happy with simple negative valued cases like:

```python
>>> import numpy as np
>>> x0 = np.longdouble(-1.0)
>>> x1 = np.nextafter(x0, np.inf)
>>> x2 = np.nextafter(x1, np.inf)
>>> x0, x1, x2
(-1.0, -0.99999999999999999995, -0.9999999999999999999)
>>> from numpy.testing import assert_array_almost_equal_nulp
>>> assert_array_almost_equal_nulp(x0, x1)
>>> assert_array_almost_equal_nulp(x0, x2)
```

```
>       with pytest.raises(AssertionError, match="1 ULP"):
E       Failed: DID NOT RAISE <class 'AssertionError'>
E       Falsifying example: test_gh_21520(
E           val=array([-1.], dtype=float128),
E           self=<numpy.testing.tests.test_utils.TestArrayAlmostEqualNulp at 0x7f80b4180650>,
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.60.0', b'AAAAAAEAAAAAAAAAAQ==') as a decorator on your test case

self       = <numpy.testing.tests.test_utils.TestArrayAlmostEqualNulp object at 0x7f80b4180650>
val        = array([-1.], dtype=float128)
x0         = array([-1.], dtype=float128)
x1         = array([-1.], dtype=float128)
x2         = array([-1.], dtype=float128)
```

Based on the CI, it looks like `hypothesis` may need some convincing to use a width of `128` in some cases.. not sure if we'd need to disable that on some platforms (I was hoping the `hypothesis` strategy would auto-adjust based on the size in bits of `longdouble` on a given platform). It looks like only certain parts of `hypothesis` tolerate 128 bits to begin with, but at least I can see that locally with 128 bit `longdouble` it is at least possible to flush through at that width.

seberg commented:
If you want to make some manual tests, how about just using numbers like `np.finfo(...).max` and calaling `np.nextafter()` `n` times from there to test against?  Or just random large values (e.g. `np.sqrt(np.finfo(...).max)`?).
And I guess the `np.finfo(...).max` to `np.inf` step needs to be tested explicitly.

Testing longdouble numbers might be hard using hypothesis, unless maybe you generate a 128bit integer and the exponent separately?

tylerjereddy commented:
Well, for negative values, we're not really using ULP anymore because the function uses absolute value. Even if you switch to `float64` it gets confusing:

```python
import numpy as np
from numpy.testing import assert_array_almost_equal_nulp

x0 = np.float64([-1.0])
x1 = np.nextafter(x0, np.inf)
x2 = np.nextafter(x1, np.inf)

# doesn't raise
assert_array_almost_equal_nulp(x0, x2, nulp=1)
```

That said, the docs do indeed specify that we're talking about "magnitude:"

> An assertion is raised if the following condition is not met:

> abs(x - y) <= nulp * spacing(maximum(abs(x), abs(y)))

So, would we agree that it is fair game for me to only require the raises to happen on the right side of the "textbook ULP lines" then?

![image](https://user-images.githubusercontent.com/7903078/213944494-1ac00b66-2a7f-4764-ac76-575b559b9968.png)

That may provide me with some interesting shortcuts. In the meanwhile, I'll double check some behavior on `assert_array_max_ulp`, which is closer to what I would "intuitively" want to use in most cases I would think, based on the description. Wouldn't it be a weird use case to have negative values and want to use `assert_array_almost_equal_nulp`? Then you're basically saying "I want to use postive-side ULPs for negative numbers?" Slightly more relaxed comparison in a weird way?

seberg commented:
Ah, I suppose I do agree with that, and the way the docs are written we always use the biggest spacing to define the nulps difference.
I do wonder what we actually want in practice, this is a bit more confusing than I thought, because I expected it to be the actual `nextafter` steps between numbers (at least for small differences).

tylerjereddy commented:
Hmm, if we actually intend to deprecate `longdouble` per https://github.com/numpy/numpy/pull/23069#issuecomment-1409418063 I'm wondering if I should continue here. I guess if you agree with Ralf this can be closed if it isn't worth it.

seberg commented:
I don't have a strong opinion either way.  Some people will be hurt by removal, but not super many and some of them have said they could work on a replacement living outside of NumPy.  A few other users may use it just for very limited operations, where a replacement might be nice, but maybe not necessary (i.e. a special stable sum function, or quad precision because speed doesn't matter, or...).
